django/db/models/fields/__init__.py

```diff
@@ -242,31 +242,55 @@ def _check_field_name(self):
             return []
 
     def _check_choices(self):
-        if self.choices:
-            if isinstance(self.choices, str) or not is_iterable(self.choices):
-                return [
-                    checks.Error(
-                        "'choices' must be an iterable (e.g., a list or tuple).",
-                        obj=self,
-                        id='fields.E004',
-                    )
-                ]
-            elif any(isinstance(choice, str) or
-                     not is_iterable(choice) or len(choice) != 2
-                     for choice in self.choices):
-                return [
-                    checks.Error(
-                        "'choices' must be an iterable containing "
-                        "(actual value, human readable name) tuples.",
-                        obj=self,
-                        id='fields.E005',
-                    )
-                ]
-            else:
-                return []
+        if not self.choices:
+            return []
+
+        def _is_terminal(value):
+            return isinstance(value, str) or not is_iterable(value)
+
+        if _is_terminal(self.choices):
+            return [
+                checks.Error(
+                    "'choices' must be an iterable (e.g., a list or tuple).",
+                    obj=self,
+                    id='fields.E004',
+                )
+            ]
+
+        # Expect [group_name, [value, display]]
+        for choices_group in self.choices:
+            try:
+                group_name, group_choices = choices_group
+                try:
+                    if not all(
+                        _is_terminal(value) and _is_terminal(human_name)
+                        for value, human_name in group_choices
+                    ):
+                        break
+                except (TypeError, ValueError):
+                    # No groups, choices in the form [value, display]
+                    value, human_name = group_name, group_choices
+                    if not _is_terminal(value) or not _is_terminal(human_name):
```

Comment:
There's a bunch of boolean logic going on, so forgive me if I'm reading it through wrong - `if not _is_terminal(value)` is that not going to raise a system check error for non-string *values*?

Reply:
It's going to raise a system check error if `value` is not a terminal. A terminal is defined as something that's either a string or not iterable (mainly: `int`,`float`, ...). So the system check error will only be raised if value is an iterable.

The term of `terminal` is borrowed from the [terminal and non-terminal symbols](https://en.wikipedia.org/wiki/Terminal_and_nonterminal_symbols), but I can change it if another name makes it clearer?

Also, `break` here might be a bit confusing, it actually forces the loop to exit without executing the `else` clause, raising an error.

Hope that helps?

Reply:
> [...] So the system check error will only be raised if value is an iterable.

Yeah so I was basically trying to figure out whether integer values (ie: `IntegerField() `with `choices`) etc were still supported, and the combo of boolean logics got me confused. All I've done is add noise, sorry.

Reply:
No worries, I am happy to answer any question. Thank you for taking time to review this PR!